### PR TITLE
simplify reader and enable use_shared_memory

### PR DESCRIPTION
**Pull request by heavengate**
**simplify reader and enable use_shared_memory**
- Simplify `DataReader` implement since DataLoader support `dict` format(https://github.com/PaddlePaddle/Paddle/pull/31481)
- enable `use_shared_memory` if `/dev/shm` size is larger than 1G

### Speed test

YOLOv3 training speed improved, this PR has nearly no influence on Mask-RCNN/SSD training speed

**Testing enviorments**
- 4 * Tesla V100(16G)
- CUDA 10.2
- cuDNN 7.6
- shared memory size: 4G

| Model | batch_size | worker_num | develop speed | this PR speed | speed up ratio |
| :--: | :--:|  :--: |  :--: |  :--: |  :--: | 
| YOLOv3-DarkNet53 | 4 * 8 | 4 * 2 | 14.27 samples/s | 22.35 samples/s | 56.6% |
| Mask RCNN-R50 | 4 * 1 | 4 * 2 | 2.52 samples/s | 2.51 samples/s | - |
| SSD-VGG16 | 4 * 8 | 4 * 2 | 22.36 samples/s | 23.41 samples | - |

### CODE REVIEW

**Review by wangxinxin08 — APPROVED** [review]
LGTM
